worker-workflow/src/main/resources/workflow-control.js

```diff
@@ -193,6 +193,68 @@ function applyActionDetails(document, actionDetails, terminateOnFailure) {
     }
 }
 
+function onAfterProcessDocument(e) {
+    if (!e.application.getInputMessageProcessor().getProcessSubdocumentsSeparately() && e.rootDocument.hasSubdocuments()) {
+        processSubdocumentFailures(e.rootDocument.getSubdocuments());
+    }
+    processFailures(e.document);
+}
+
+function processSubdocumentFailures(subdocuments) {
+    for each(var subdoc in subdocuments) {
+        if (subdoc.hasSubdocuments()) {
+            processSubdocumentFailures(subdoc.getSubdocuments());
+        } else {
+            processFailures(subdoc);
```

Comment:
This logic doesn't seem right.  You appear to be skipping the parent documents.

Reply:
The parent is processed in row 200

Reply:
That's the root parent.  It is the interim parents that you appear to be skipping.

Reply:
Added a function to deal with it

Reply:
Let's look at this together tomorrow - perhaps I'm wrong but the new logic seems overly complicated and I'm struggling to follow it to confirm that it is correct.

Reply:
As discussed, something like this;

    function processFailuresInHierarchy(document)
    {
        processFailures(document);

        for each(var subdocument in document.getSubdocuments()) {
            processFailuresInHierarchy(subdocument);
        }
    }

Reply:
This is still unnecessarily complex.  You're still using `.hasSubdocuments()`, which is leading you to adding extra unnecessary code branches.  Recursive functions should be kept as simple as possible.